src/shared/metadata/selectors.js

```diff
@@ -412,22 +416,46 @@ const resolveCategoryOptionIds = (categories, categoryOptions) => {
 
 export const getCategoriesWithOptionsWithinPeriod = createCachedSelector(
     (metadata) => metadata,
-    (_, dataSetId) => dataSetId,
+    getDataSetById,
     (_, __, periodId) => periodId,
-    (metadata, dataSetId, periodId) => {
-        if (!dataSetId || !periodId) {
+    (metadata, dataSet, periodId) => {
+        if (!dataSet?.id || !periodId) {
             return []
         }
 
-        const relevantCategories = getCategoriesByDataSetId(metadata, dataSetId)
+        const relevantCategories = getCategoriesByDataSetId(
+            metadata,
+            dataSet?.id
+        )
+
         const categoryOptions = getCategoryOptions(metadata)
         const relevantCategoriesWithOptions = resolveCategoryOptionIds(
             relevantCategories,
             categoryOptions
         )
         const period = parsePeriodId(periodId)
         const periodStartDate = new Date(period.startDate)
-        const periodEndDate = new Date(period.endDate)
+
+        // periodEndDate is up to following start date
+        let periodEndDate = addFullPeriodTimeToDate(
```

Comment:
Forgive my ignorance, I'm not too familiar with the date stuff here. But what is the difference between this (and the `periodEndDate.setDate(periodEndDate.getDate() - 1)` below), and using the previous `new Date(period.endDate)`.

Reply:
@Birkbjo: so previously we were just taking the endDate for the period, which was easy because we could just create a date object from the endDate associated with the period, so we create a date object for April 2022 from `new Date('2022-04-30')`, for example.

When we need to reduce the date however, we can get into a situation where just reducing the endDate itself will cause problems. For example, if we decrement the month on `2022-04-30`, we get `2022-03-30`, which is now not the end of the period March 2022. For that reason, in this code, I'm initiating the endDate as the next period's start date (e.g. instead of `new Date('2022-04-30')`, I used `new Date('2022-05-01')`, then decrement the periods, e.g. decrement 1 month to get `2022-04-01`, then subtract a day to get `2022-03-31` (i.e. we get to the endDate for the previous period).

Agree (in light of the comment below), that it would be better to move this logic elsewhere

Reply:
Ah, of course. That makes perfect sense!